File: src/lambda/dms-switch/dms-pre.ts
import { CloudFormationClient } from '@aws-sdk/client-cloudformation';
import { DatabaseMigrationServiceClient, StopReplicationTaskCommand } from '@aws-sdk/client-database-migration-service';
import type {
  CloudFormationCustomResourceEvent,
  CloudFormationCustomResourceFailedResponse,
  CloudFormationCustomResourceSuccessResponse,
} from 'aws-lambda';
import { getDmsTask } from './utils/get-dms-task';
import { hasDmsChanges } from './utils/has-dms-changes';
import { getDmsStatus } from './utils/get-dms-status';
import { waitForDmsStatus } from './utils/wait-for-dms-status';

const dms = new DatabaseMigrationServiceClient({});
const cf = new CloudFormationClient({});
let ReplicationTaskArn: string;

export const handler = async (
  event: CloudFormationCustomResourceEvent,
): Promise<CloudFormationCustomResourceSuccessResponse | CloudFormationCustomResourceFailedResponse> => {
  try {
    const StackName = `${process.env.STACK_NAME}`;
    if (!ReplicationTaskArn) {
      ReplicationTaskArn = await getDmsTask({ cf, StackName });
    }
    const status = await getDmsStatus({ dms, ReplicationTaskArn });
    if (status === 'running') {
      if (event.RequestType === 'Delete' || await hasDmsChanges({ cf, StackName })) {
        // pause task
        const stopCmd = new StopReplicationTaskCommand({
          ReplicationTaskArn,
        });
        await dms.send(stopCmd);
        // wait for task to be fully paused
        await waitForDmsStatus({ dms, ReplicationTaskArn, targetStatus: 'stopped' });
      }
    }
    return { ...event, PhysicalResourceId: 'pre-dms', Status: 'SUCCESS' };
  } catch (e) {
    console.error(`Failed!`, e);
    return { ...event, PhysicalResourceId: 'pre-dms', Reason: (e as Error).message, Status: 'FAILED' };
  }
};

File: src/lambda/dms-switch-py/dms_pre.py
import os
import boto3
from utils.get_dms_task import get_dms_task
from utils.get_dms_task_status import get_dms_task_status
from utils.has_dms_changes import has_dms_changes
from utils.wait_for_dms_status import wait_for_dms_status


dms = boto3.client('dms')
cf = boto3.client('cloudformation')


def handler(event, context):
    try:
        stack_name = os.environ.get('STACK_NAME')
        replication_task_arn = get_dms_task(cf, stack_name)
        status = get_dms_task_status(replication_task_arn)
        if status == 'running':
            if event['RequestType'] == 'Delete' or has_dms_changes(cf, stack_name):
                stop_cmd = {
                    'ReplicationTaskArn': replication_task_arn
                }
                dms.stop_replication_task(**stop_cmd)
                wait_for_dms_status(replication_task_arn, 'stopped')
        return {
            'PhysicalResourceId': 'pre-dms',
            'Status': 'SUCCESS'
        }
    except Exception as e:
        print('Failed!', str(e))
        return {
            'PhysicalResourceId': 'pre-dms',
            'Reason': str(e),
            'Status': 'FAILED'
        }

